src/api/index.md

```diff
@@ -0,0 +1,277 @@
+# Gutenberg API Prototype
+
+This is proposal for a block API for the new WordPress editor!
+
+Joen Asmussen pointed out that the design of the new editor would be
+"block first". This sunk in and spurred this adventure into creating an API that
+itself is just a single self referencing "block" that can create other blocks
+out of itself.
+
+## Overview
+
+### How does this thing work
+
+While trying to create a block that could make other blocks, which could make
+other blocks, with a nice syntax, I settled on the pattern seen in blockType.js
+et al. Basically, you have a function, that takes a function of properties and
+returns a new reference to a function assigned those properties. This is then
+partially applied to return a new function that takes properties. Each time the
+function is called a new reference is created to enable the ability for new
+entities to be created from a primitive version of itself carrying along its own
+properties and at the beginning starting with none.
+
+### Quick intro
+
+`blockType()`, `block()`, `control()`, and `controlType()`, all serve as an independent
+tool for building larger pieces. Each are identical other than their name. They
+all work from the same fundamental pattern. Let's look at `blockType`, and
+`controlType`. These would serve as a way to create factories for `block()` and
+control(). A good idea might be to have a `create()` method on a type, which would
+create the actual block, or control. Then each block or control would have a
+`render()` method which would create an actual representation of that block or
+control. `render()` could be used to return React components, DOM, vDOM, HTML,
+TinyMCE, whatever is chosen. Let's look at creating a `blockType`.
+
+### Creating a textBlockType
+
+```js
+// textBlock would be the block
+import textBlock from '../blocks/textBlock'
+import blockType from './blockType'
+
+const textBlockTypeProperties = {
+	type: 'wp-text',
+	title: 'Text',
+	controls: [
+		[
+			controlType( {
+				type: 'left-align',
```

Comment:
> What purpose is type serving? Is this a unique identifier for the control? If so, is it unique across all blocks (allowing controls to be reused)?

Yes

>  Could this be a key on the controls object?

You can do that, but I would not suggest it.  If you have a list of controls it should be a list of control objects.  Not a key value storage.  Obviously that is preference, but overtime I have drifted to the idea that if it is a list of something it should indeed be an array.  This makes iterating over it a lot easier.  If we wanted to iterate over the types of controls by changing it to an object we would need to iterate over object keys and match that to the object.  Instead of just looping through the controls.